src/modules/rlm_perl/rlm_perl.c

```diff
@@ -394,6 +394,123 @@ static int perl_vblist_to_av(AV *av, fr_value_box_list_t *head) {
 	return 0;
 }
 
+/** Parse a Perl SV and create value boxes, appending to a list
+ *
+ * For parsing values passed back from a Perl subroutine
+ *
+ * When hashes are returned, first the key is added as a value box then the value
+ *
+ * @param[in] ctx	to allocate boxes in.
+ * @param[out] list	to append value boxes to.
+ * @param[in] request	being handled - only used for debug messages
+ * @param[in] sv	to parse
+ * @return
+ * 	- 0 on success
+ * 	- -1 on failure
+ */
+static int perl_sv_to_vblist(TALLOC_CTX *ctx, fr_value_box_list_t *list, request_t *request, SV *sv) {
+	fr_value_box_t	*vb = NULL;
+	char		*tmp;
+	STRLEN		len;
+	AV		*av;
+	HV		*hv;
+	I32		sv_len, i;
+	int		type;
+
+	type = SvTYPE(sv);
+
+	switch (type) {
+	case SVt_IV:
+	/*	Integer or Reference */
+		DEBUG3("Integer or Reference returned");
+		if (SvROK(sv)) {
+			if (perl_sv_to_vblist(ctx, list, request, SvRV(sv)) < 0) return -1;
+			break;
+		}
+		MEM(vb = fr_value_box_alloc(ctx, FR_TYPE_INT32, NULL, false));
```

Comment:
DEBUG3("Integer")

Reply:
I think we can actually get the taint status from perl as well for all of these.